File: src/utils/parse-size.test.ts
import { describe, expect, it } from "vitest";

import { parseSize } from "./parse-size.js";

describe("parseSize", () => {
  it("parses bytes", () => {
    expect(parseSize("1B")).toBe(1);
    expect(parseSize("100B")).toBe(100);
  });

  it("parses kilobytes", () => {
    expect(parseSize("1K")).toBe(1024);
    expect(parseSize("2K")).toBe(2048);
  });

  it("parses megabytes", () => {
    expect(parseSize("1M")).toBe(1048576);
    expect(parseSize("2M")).toBe(2097152);
  });

  it("parses gigabytes", () => {
    expect(parseSize("1G")).toBe(1073741824);
  });

  it("is case-insensitive", () => {
    expect(parseSize("1m")).toBe(1048576);
    expect(parseSize("1k")).toBe(1024);
  });

  it("handles decimal values", () => {
    expect(parseSize("1.5M")).toBe(Math.floor(1.5 * 1024 * 1024));
  });

  it("trims whitespace", () => {
    expect(parseSize("  1M  ")).toBe(1048576);
  });

  it("throws on invalid format", () => {
    expect(() => parseSize("abc")).toThrow("Invalid size format");
    expect(() => parseSize("")).toThrow("Invalid size format");
    expect(() => parseSize("1X")).toThrow("Invalid size format");
  });
});

File: internal/util/parse_size_test.go
package util

import "testing"

func TestParseSize(t *testing.T) {
	tests := []struct {
		name    string
		input   string
		want    int
		wantErr bool
	}{
		{name: "bytes", input: "1B", want: 1},
		{name: "kilobytes", input: "2K", want: 2048},
		{name: "megabytes", input: "1M", want: 1024 * 1024},
		{name: "gigabytes", input: "1G", want: 1024 * 1024 * 1024},
		{name: "case insensitive", input: "1m", want: 1024 * 1024},
		{name: "decimal", input: "1.5M", want: int(1.5 * 1024 * 1024)},
		{name: "trim whitespace", input: "  1M  ", want: 1024 * 1024},
		{name: "invalid", input: "abc", wantErr: true},
		{name: "empty", input: "", wantErr: true},
		{name: "unknown unit", input: "1X", wantErr: true},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got, err := ParseSize(tt.input)
			if tt.wantErr {
				if err == nil {
					t.Fatal("ParseSize returned nil error")
				}
				return
			}

			if err != nil {
				t.Fatalf("ParseSize returned error: %v", err)
			}
			if got != tt.want {
				t.Fatalf("ParseSize(%q) = %d, want %d", tt.input, got, tt.want)
			}
		})
	}
}
